obsreport/obsreport_scraper.go

```diff
@@ -0,0 +1,117 @@
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package obsreport
+
+import (
+	"context"
+
+	"go.opencensus.io/stats"
+	"go.opencensus.io/tag"
+	"go.opencensus.io/trace"
+
+	"go.opentelemetry.io/collector/config/configmodels"
+	"go.opentelemetry.io/collector/consumer/consumererror"
+)
+
+const (
+	// Key used to identify scrapers in metrics and traces.
+	ScraperKey = "scraper"
+
+	// Key used to identify metric points scraped by the Collector.
+	ScrapedMetricPointsKey = "scraped_metric_points"
+	// Key used to identify metric points errored (ie.: unable to be
+	// scraped) by the Collector.
+	ErroredMetricPointsKey = "errored_metric_points"
+)
+
+var (
+	tagKeyScraper, _ = tag.NewKey(ScraperKey)
+
+	scraperPrefix                 = ScraperKey + nameSep
+	scraperMetricsOperationSuffix = nameSep + "MetricsScraped"
+
+	mScraperScrapedMetricPoints = stats.Int64(
+		scraperPrefix+ScrapedMetricPointsKey,
+		"Number of metric points successfully scraped.",
+		stats.UnitDimensionless)
+	mScraperErroredMetricPoints = stats.Int64(
+		scraperPrefix+ErroredMetricPointsKey,
+		"Number of metric points that were unable to be scraped.",
+		stats.UnitDimensionless)
+)
+
+// ScraperContext adds the keys used when recording observability metrics to
+// the given context returning the newly created context. This context should
+// be used in related calls to the obsreport functions so metrics are properly
+// recorded.
+func ScraperContext(
+	ctx context.Context,
+	scraper string,
+) context.Context {
+	ctx, _ = tag.New(ctx, tag.Upsert(tagKeyScraper, scraper, tag.WithTTL(tag.TTLNoPropagation)))
+	return ctx
+}
+
+// StartMetricsScrapeOp is called when a scrape operation is started. The
+// returned context should be used in other calls to the obsreport functions
+// dealing with the same scrape operation.
+func StartMetricsScrapeOp(
+	scraperCtx context.Context,
+	scraper string,
+) context.Context {
+	spanName := scraperPrefix + scraper + scraperMetricsOperationSuffix
+	ctx, _ := trace.StartSpan(scraperCtx, spanName)
+	return ctx
+}
+
+// EndMetricsScrapeOp completes the scrape operation that was started with
+// StartMetricsScrapeOp.
+func EndMetricsScrapeOp(
+	scraperCtx context.Context,
+	numScrapedMetrics int,
+	err error,
+) {
+	numErroredMetrics := 0
+	if err != nil {
+		if partialErr, isPartial := err.(consumererror.PartialScrapeError); isPartial {
+			numErroredMetrics = partialErr.Failed
+		} else {
+			numErroredMetrics = numScrapedMetrics
+			numScrapedMetrics = 0
+		}
+	}
+
+	span := trace.FromContext(scraperCtx)
```

Comment:
Is this span guaranteed to always be the one created by StartMetricsScrapeOp?  Should we explicitly return the span from StartMetricsScrapeOp, and accept it as an argument in EndMetricsScrapeOp instead?

Or alternatively, make StartMetricsScrapeOp return an additional struct which contains the span and has an `End()` or `RecordResult()` function.

```golang
func StartMetricsScrapeOp(...) (context.Context, *MetricsScrapeOp) {
    ...
    ctx, span := trace.StartSpan(scraperCtx, spanName)
    return ctx, &MetricsScrapeOp{ctx: ctx, span: span}
}

// or you could instead make this unexported, and add an interface.
type MetricsScrapeOp struct {
    ctx context.Context
    span trace.Span
}

func (m *MetricsScrapeOp) End(numScrapedMetrics int, err error) {
   // use m.ctx and m.span
}
```

Reply:
That's a very good point. I like your idea of returning a struct with an `End()` function.

I don't want to change this here though as all the other `obsreport_*.go` files follow this same pattern. Worth doing this in a followup PR though.